### DISCUSSION TITLE
Please give us feedback and improvement points!

### DISCUSSION THREAD

**Discussion by PatrickAlphaC** [Cyfrin/foundry-full-course-cu#2021] [General]
In this thread, please leave feedback on your favorite parts, and least favorite parts of the course!

This way, we can keep updating so we can make Cyfrin Updraft easier and easier for you to learn. 

CURRENT FEEDBACK POINTS:
- Quizzes 
- More challenges

**Comment by KeySkills** [comment]
In Advanced Foundry Course, Section 3 - Develop Defi Protocol [Lesson 16: DSCEngine advanced testing](https://updraft.cyfrin.io/courses/advanced-foundry/develop-defi-protocol/defi-protocols-advanced-testings-testing) and [Lesson 17: Write fuzz tests](https://updraft.cyfrin.io/courses/advanced-foundry/develop-defi-protocol/defi-writing-fuzz-tests) and  contain the same video.

**Reply by PatrickAlphaC** [reply]
Thanks! We will fix.

**Reply by HonourJAH** [reply]
Front-end and web3.js or ether.js course would be appreciated.

**Reply by EngrPips** [reply]
There is something in the pipe for front end I believe

**Reply by KillerwhaleFx** [reply]
I love it here

**Reply by EngrPips** [reply]
> I love it here

We love to have you.

**Reply by 0xAnalystt** [reply]
i just finished blockchain overview, L1s, L2s and rollups, my gosh it was really worth it. taking a break now tho.

**Reply by EngrPips** [reply]
> i just finished blockchain overview, L1s, L2s and rollups, my gosh it was really worth it. taking a break now tho.

I am glad you  enjoyed it.

**Comment by Aaryan-Urunkar** [comment]
First of all, a supreme effort and a mega thanks to the Cyrfin team and @PatrickAlphaC  sir for creating this course! I just had a few pieces of feedback: 
1. The ERC-20 token and the NFT part seemed a bit rushed, which I don't know why put me off from finishing video 2(or even touching the course) for nearly a week. I don't know if it was just me or anybody else as well, but yeah, thought of putting it down here.

2. Secondly, I had migrated to this course after watching nearly 8 hours of the Hardhat one from freecodecamp which is outdated and hence I found it difficult to follow that one. It would be VERY VERY appreciated if the team could do a similar frontend course. Even if it isn't an entire mammoth course with the new updated Hardhat but just the frontend part for some of the existing foundry projects from this course, I would appreciate it, and I think many others would too!


Thankyou again!!!

**Reply by PatrickAlphaC** [reply]
> ERC-20 token and the NFT part seemed a bit rushed

Thanks for this feedback! Which parts would you have wished we went over deeper / slower so it would feel less rushed? 

> a similar frontend course.

This year ;)

**Reply by Aaryan-Urunkar** [reply]
The parts I feel you should have dived deeper into is definitely during the ERC20 intro section. I left that section feeling doubtful if I understood it or not but yeah, during DSC, I had to google a lot of stuff abt ERC-20 which I previously was unclear about. Hence I feel that the ERC-20 section can slightly be improved so that by DSC everyone has a better idea abt what ERC-20s are.

Very hyped up for the frontend course sir!! Already know it's gonna be a bombshell as soon as it is out! In the meanwhile, just gonna build up project from whatever I have learned so far.

EDIT: Thanks to this course I ended up with my first ever answer on stack overflow lol! (https://stackoverflow.com/a/78801142)

**Comment by bibhashdash** [comment]
Thanks for the course it's really good. I was wondering, is there a way to flag (or bookmark) a specific video so that we can easily come back to it later? I know we can leave it 'unchecked' but perhaps a quicker way would be to maybe filter and find all the bookmarked ones. 
Anyway, just a thought. Great work on the course and looking forward to continuing my journey in web3.

**Reply by PatrickAlphaC** [reply]
Oooo interesting. Good feature request!

**Comment by SiewLin-AIDeaBuddy** [comment]
Thanks a lot for the excellent educative course materials. The option to learn sectioned videos in Updraft instead of a combined long hours video is greatly appreciated. Would like to give generic improvement feedbacks as below:

1. To include feature where user can easily take notes at any timestamp of the course tutorial videos  in order to further deep dive on certain topics at later stage or to jot down topics that need attention. User can also easily view/search those notes made at various timestamps to revise the topics again instead of searching and rewatching the entire video to find the one that is relevant to revise.

2. It would be very helpful if after each new topic, a relevant past audit case study with solution or actual hacked protocols walk through is included at the end / or at the right time junction within the course to further enhance the understanding and applicability of the topic in real case. 

3. Similar to 2 above, at the end of several topics/chapters/modules, if there is a relevant section in first flight / audits that closer to the taught topics, the inclusion of those info can be added apart from any planned challenges/quizzes. This is especially useful for those who just started in the field to get closer to something that they can attempt to after the lesson to test out their understanding of the topic rather than hunting around the various sources to find out that the real test/audit is far beyond from what they have learned at that point, and couldn’t really apply due to lots of missing pieces of associated knowledge. Perhaps can utilize / modify past first flight/audit to cater the learning needs/tests for targeted topics, this makes the learning more effective as it is closer to actual applications.

**Reply by EngrPips** [reply]
Interesting feedback.

**Reply by Equious** [reply]
@Eversmile12 Some food for thought.

**Comment by mpuj99** [comment]
Thanks for this course Patrick, it was amazing. Really well explained.
Two things I noticed during these course:
1. In section 3 - Develop DEFI protocol, in [Lesson 20: Create the collateral redeem handler](https://updraft.cyfrin.io/courses/advanced-foundry/develop-defi-protocol/defi-handler-redeeming-collateral?lesson_format=video), when you create the line to get the collateral of a user:
```solidity
uint256 maxAmountCollateralToRedeem = engine.getCollateralDepositedOfUser(sender, address(collateral));
```
The function it has this order of arguments:
```solidity
function getCollateralDepositedOfUser(address user, address token) external view returns(uint256)
```
First is the address of the user and then the address of the token, in the video you put first the token and then the user:
```solidity
uint256 maxAmountCollateralToRedeem = engine.getCollateralDepositedOfUser(address(collateral), sender);
```
I suppose it doens't fail because both are addresses, I looked at the logs and when we console ```maxAmountCollateralToRedeem``` is equal to zero. So never fails, if you put it good the arguments it fails with the assertion failed: 
```(totalWethValue + totalWbtcValue) >= totalSupply``` in ```InvariantTest.t.sol```
I tried to fix it but it was hard, because I think the problem was that we were depositing collateral, minting and redeeming but in that lesson we weren't burning so, I fixed only for 5 invariant executions, creating redeeming and burning in the same function, on the 6th execution found some value that cracked the healthFactor, so I gave up because I spent too much time on creating the function to redeem and burn at the same time. So just telling you, maybe you can fix it in a new video with the arguments good positioned.
But overall, just want to say that the DEFI protocol section was my favourite section of Advanced Foundry, I learned a lot. 

2. Is a silly thing but, in the last section: Security, in the [lesson 3: Top security tools](https://updraft.cyfrin.io/courses/advanced-foundry/security/top-smart-contract-security-tools?lesson_format=video), from the minute 9:23 till 9:34 is all black, is nothing but just saying maybe you didn't notice it.
Thank you very much, amazing course. With eager to start hitting the security and auditing course!!!!

Regards,

**Reply by EngrPips** [reply]
Awesome Feedback, I trust he will reply to this as soon as he is available.

**Reply by PatrickAlphaC** [reply]
So sorry I missed this! Thanks for the feedback here :) We will fix!

**Comment by thxpikachu** [comment]
In Blockchain Basics, Section 6 - The Purpose of Smart Contracts, the video cuts off a bit too early. The sentence was supposed to end with "and wash out the products that aren't actually contributing to Web3," 
but on Cyfrin Updraft, it abruptly stops at  "aren't actually contribut-".

It's not a big deal, but my perfectionist brain can't bear it. :)

**Reply by EngrPips** [reply]
Thanks for the feedback, This will be attended to appropriately as soon as possible.

**Reply by Equious** [reply]
FYI @PatrickAlphaC 🙏

**Comment by MattSWeb3** [comment]
In "Blockchain Basics **11. Quiz: Blockchain Basics Mid Section Recap**" Question 4 "What is the role of the gas price in ETH?" refers to the next video of the course "**12. Introduction to Gas**". Shouldn't the videos be the other way around? Since the quiz asks about something we didn't see up to that point.

**Reply by EngrPips** [reply]
Thanks for the feedback, This will be attended to appropriately as soon as possible.

**Comment by Dennnnny** [comment]
hey, I have a little question about quiz below: 
In [Foundry Fundamentals](https://updraft.cyfrin.io/courses/foundry) Section4 Smart Contract Lottery  11. Quiz: Smart Contract Lottery Intro 
number 4 about `How can you use the modulo operator (%) to determine if a number is even or odd?`

I think `If n % 2 != 1, the number is even. Otherwise, it's odd.` this option is also correct (?) 
because if a number % 2, is either 0 or 1

<img width="904" alt="image" src="https://github.com/user-attachments/assets/a874c186-d6a5-41f7-9635-5d01ccaeface">

**Reply by EngrPips** [reply]
I think you have a point, This will be looked into as soon as possible.

**Reply by felixoder** [reply]
The logic you mentioned is partially correct, but not always reliable due to how modulus works with negative numbers. Here's why:

1.  For positive numbers, n % 2 != 1 would work, because positive odd numbers will have a remainder of 1 when divided by 2 {Two}.
2.   However, for negative odd numbers, the result of n % 2 is not necessarily 1. For example, -3 % 2 gives -1, not 1. So the check n % 2 != 1 would fail for negative odd numbers.

The safer way is to simply check n % 2 == 0 for even numbers (both positive and negative), and if not, the number is odd.

**Comment by MattSWeb3** [comment]
In **Blockchain Basics** course the written lesson on **16. Blockchain Overview** has a small mistake in:
"A `chain selection rule` is implemented as a means to determine which blockchain is the _real_ blockchain. Bitcoin (and prior to the merge, Ethereum), both use something called `**Nakomoto** Consensus`. This is a combination of Proof of Work (Etherum has since switched to Proof of Stake) and the `longest chain rule`."
It should be written "Nakamoto". 

Thanks for the amazing work you guys have been doing. I'm loving the course!!

**Reply by EngrPips** [reply]
Thanks for pointing this out and this will be attended to as soon as possible.

**Reply by Equious** [reply]
This has been updated! In future you can definitely submit PRs for the written lessons to the dev branch on this repo: https://github.com/Cyfrin/Updraft

**Comment by khandelwalmoksh** [comment]
In "New discussion" section, "Preview" isn't working properly
my code:
 ```function testOwnerIsMsgSender() public {
        assertEq(fundMe.i_owner(), address(this));
    }```

my code ``` function testOwnerIsMsgSender() public {
        assertEq(fundMe.i_owner(), address(this));
} ```
    
![Screenshot (18)](https://github.com/user-attachments/assets/5c871d3f-68f9-46e0-9367-30255c426a95)
![Screenshot (19)](https://github.com/user-attachments/assets/1e3140e4-5800-46ff-b9d3-7043b5add0ec)


U can clearly see the difference, the 1st code is incomplete whereas, 2nd is complete

**Reply by EngrPips** [reply]
I am not sure I understand the issue you are pointing out here.

**Reply by khandelwalmoksh** [reply]
![Screenshot (18)](https://github.com/user-attachments/assets/207546d9-2e6f-47db-919f-6423ce12909b)
![Screenshot (19)](https://github.com/user-attachments/assets/89ab8927-6cff-4035-b8ef-3541fefff4dd)

these must help you 
U can clearly see the difference, the 1st code is incomplete whereas, 2nd is complete

**Reply by EngrPips** [reply]
I see, It will be looked into. thanks for the contribution.

**Reply by PatrickAlphaC** [reply]
Place the ``` on their own line, like this:

````
```
hi!
```
````

**Reply by chandalalaguidance-beep** [reply]
What kind of software among the Microsoft office package can you use to make your EIPS and CRC token
Where can I find the lesson about web designing

**Reply by EngrPips** [reply]
> What kind of software among the Microsoft office package can you use to make your EIPS and CRC token Where can I find the lesson about web designing

If you search the web, you will find a lesson about web design.

**Comment by khandelwalmoksh** [comment]
"IF YOU CAN" provide Patrick's codes used in Foundry Fundamentals, it would be very beneficial. Using the Solidity codes and the ones given in the repo, are giving countless errors. Even Phind ai can't help. And @EngrPips & others has to go through every line of the code provided to figure out what's wrong. This process eats up 2-3 days, thereby wasting the precious time of both, the questioner and solver. Thank you

**Reply by EngrPips** [reply]
Hello @khandelwalmoksh, I think you should not bother about the final code of each course but instead code along with Patrick, and like that, you won't run into too many problems, what do you think?

**Reply by PatrickAlphaC** [reply]
We do provide the code for each! Are you unable to find it?

**Comment by chainflash** [comment]
lesson 6 part 3 to 7 is a bit difficult for windows users

**Reply by EngrPips** [reply]
Thanks for pointing this out and this will be attended to as soon as possible.

**Comment by dvinubius** [comment]
Wrong answer for question 1 in the **Quiz Fund Me Section Recap**

Choice 3 is not entirely accurate. Constants cannot be assigned at any point, but must be assigned with an explicit value in the source code, so that the value is known at compile time. They are being specified at compile time and effectively set in storage on deployment.

Choice 4 is arguably correct though. Immutables can be set with one value when declared, and then **updated** when the constructor runs. Constants, on the other hand, can't ever be **updated** (which implies they were having another value before). They can only be specified once, at compile time, and then set accordingly when initialized, namely on deployment.

<img width="1070" alt="Screenshot 2024-10-26 at 22 50 13" src="https://github.com/user-attachments/assets/824765ef-47cf-412c-8842-99ce88744ba6">

**Reply by EngrPips** [reply]
Thanks for pointing this out and this will be attended to as soon as possible.

**Reply by Equious** [reply]
Updated this question with some hopefully more clear/appropriate answers. Thanks for the feedback

**Comment by uddercover** [comment]
I noticed a blank part of the video in [lesson 4, advanced foundry security section. ](https://updraft.cyfrin.io/courses/advanced-foundry/security/top-smart-contract-security-tools)
Timestamp: 9:22 - 9:34

**Reply by EngrPips** [reply]
Thanks for this feedback. This will be looked into.

**Reply by Equious** [reply]
This seems to be fixed. Thanks for the report.

**Comment by engAbody** [comment]
i had been in many other courses looking for problem to practice ,
i found that even chainlink make training question in code format at Githup 

not hard and does not require special website 

Thanks for consideration ,,

**Comment by intuitzer** [comment]
idk how but the vibe and aura of the course and instructor aka Patrick (hats off to him) just generates an extra chunk of interest in the field, it can't get better.
Thanks a lot. :)

**Reply by EngrPips** [reply]
You are welcome. It will only get better.

**Comment by MutteTejeshu** [comment]
Hi, I guess there was some mistake in the explanations of Max Priority Fee at 4:05 in Blockchain Basics - 15. Gas in depth. It was explained that Max Priority Fee is Max gas fee plus the Max Tip, but isn't that just the Max Tip?

And at 8:26 in the same video, it was said that the gas price we used was less than the max priority fee but the gas price is greater than the max priority fee.

Thanks for the great course and explanations.

**Reply by EngrPips** [reply]
Thanks for pointing out this @MutteTejeshu, @PatrickAlphaC  will attend to this as soon as possible.

**Comment by codex-scribe** [comment]
It's proving really difficult to use the faucets right now, and it cost me a lot of sepoliaETH to deploy my first smart contract.  Like .014 out of the .024 I had in my wallet (meaning I won't have enough to do a second smart contract unless I get a faucet to work again).
Maybe if Patrick knows someone in the Eth development community he can get a lot of sepolia ETH and send it out to people who request it through a thread like this.  Perhaps if the requester can show work in their github relevant to this course.  Then they could send 1 sETH to a user so they have a lot to work with for the future.  I'm not looking forward to battling these faucets every day to get a little dribble of testnet ETH.

**Reply by codex-scribe** [reply]
The google faucet is the only one I've seen that worked that you didn't have to create an account on some website for, and to use it, you have to wait 10-20 minutes after your first request, and then make a second request.

**Comment by Aashim-Limbu** [comment]
@PatrickAlphaC 
Thanks for such an amazing course . I would love to have a separate section on zkSync Rollups and discussion about everything to know about L2 and its jargons . As in the first section it self you've mentioned we'll discuss rollups in more depth later in the course.And for future would love to have course on the solana chain as well  .Thank you for such an amazing course 🙏😀.

**Reply by EngrPips** [reply]
Great suggestion!

**Comment by ismaildawoodjee** [comment]
Great courses so far, can't wait to finish 'em all. 

Is this also a place to point out bugs or mistakes? Because I found an incomplete sentence in [Solidity Smart Contract Development](https://updraft.cyfrin.io/courses/solidity), **Section 3:** Fund Me, **27. Quiz:** Fund Me Section Recap, **Question 5.** How does the fallback function differ from the receive function in Solidity?, the answer is "The receive function is specifically designed to handle Ether transfers without data and is automatically invoked when Ether." and it just stops there.

Should it be "...is automatically invoked when Ether is sent"?

**Reply by EngrPips** [reply]
Hey @ismaildawoodjee, We are glad you are enjoying the course. You should use this GitHub's `issue` tab to report that issue.

**Comment by Imjolayemi** [comment]
A very great course to start with as a beginner, but I'm having problem accessing the tenderly link located in block chain basics tutorial 12 (sending your first transaction tenderly), kindly check this out or if you have a solution to this kindly do that.

**Reply by EngrPips** [reply]
Thanks for pointing this out. But it opens fine for me. Please check that everything is fine on your end. The link  https://tenderly.co/?mtm_campaign=partner&mtm_kwd=cyfrin

![Screenshot 2025-01-18 at 04 23 36](https://github.com/user-attachments/assets/ec403a8c-4c9c-4ffc-ad4f-1e3da51e661f)

**Reply by Imjolayemi** [reply]
Yes it is working, but I think the platform is working for some specific location or ip address. Because It works on my mobile phone immediately I changed it ip address here.

Is that how I will be accessing it, could there be any solution to that?

**Reply by EngrPips** [reply]
I am unsure why you have that problem. I didn't have any issue at all accessing the website maybe you can try a vpn on your computer.

**Reply by Imjolayemi** [reply]
Okay thanks very much

**Reply by EngrPips** [reply]
Anytime, Friend.

**Comment by TamirKalimi** [comment]
Hi,
I've finished the [Blockchain Basics Course](https://updraft.cyfrin.io/courses/blockchain-basics) , and it’s been a fantastic introduction to the blockchain world. Thank you for putting this course together!  
However, I noticed a couple of areas where the course could be improved for better clarity and flow:  
1. **Repetition between Lessons 19 and 26**:  
   The ending of [Lesson 19: Blockchain Overview](https://updraft.cyfrin.io/courses/blockchain-basics/basics/how-the-blockchain-works) is identical to [Lesson 26: Congratulations](https://updraft.cyfrin.io/courses/blockchain-basics/basics/blockchian-basics-completed). 
 Differentiating the content in Lesson 26 would make it feel more like a unique and memorable conclusion to the course.  

2. **Quiz Placement and Content**:  
   In the [Lesson 22: Quiz: Blockchain Basics Recap](https://updraft.cyfrin.io/courses/blockchain-basics/basics/blockchain-basics-recap-quiz), there are questions about rollup stages that are only introduced in [Lesson 23: Rollup Stages](https://updraft.cyfrin.io/courses/blockchain-basics/basics/rollup-stages). Swapping these lessons might improve the learning flow and ensure students have the background needed before taking the quiz.

**Reply by EngrPips** [reply]
Great feedback. @PatrickAlphaC @Equious

**Reply by Equious** [reply]
Thank you, we'll have a look this week!

**Reply by PatrickAlphaC** [reply]
This is great! Thank you!

**Comment by abdulsamed1** [comment]
When will the Arabic subtitle 🇸🇦 be added?

**Reply by EngrPips** [reply]
Interesting. @Equious

**Comment by WhiteKnightK** [comment]
In the 25th lesson of the DeployRaffle course, the text portion and the video portion seem to have different content. I am not sure if it is the text course of other video content.

**Reply by EngrPips** [reply]
Awesome, This will be looked into as soon as possible. @PatrickAlphaC  @Equious sir

**Comment by CodeAtomsphere** [comment]
A feedback? My Mac just went faulty, I need a new one 🤦‍♂️

**Reply by EngrPips** [reply]
Make sure to buy minimum of 16gb ram and 512 SSD for good workflow.

**Reply by CodeAtomsphere** [reply]
Ikr, just broke 😂

**Comment by Ilikeatingvegtable123** [comment]
You could add questions in the written lessons at the end of the lesson for some courses (e.g, Foundry Fundamentals or Advanced Foundry). Some other courses have lessons that add questions at the end of it that help ingrain knowledge.

**Reply by EngrPips** [reply]
Thanks for the contribution. @PatrickAlphaC  @Equious  sir

**Comment by WhiteKnightK** [comment]
In the Advanced Foundry course, specifically in the '4. Introduction to IPFS' lesson, it seems that there are no subtitles available in Chinese. As I am not a native English speaker, I am having difficulty fully understanding the content of this video. Could you please consider adding Chinese subtitles?

**Reply by EngrPips** [reply]
Thanks for raising this concern. @PatrickAlphaC  @Equious  sir

**Comment by Equious** [comment]
Thanks, I'll have a look at generating this on Monday.

**Comment by wildjos** [comment]
Comment on **Blockchain Basics**
Section 22: Quiz: Blockchain Basics Recap
A question on "What is a Rollup Stage" has not been covered.  The answer is in the next section, 23: Rollup Stages. 

Image to illustrate:
<img width="1151" alt="Screenshot 2025-04-01 at 10 01 09" src="https://github.com/user-attachments/assets/88634046-4213-4704-85e9-d2b43d0231ca" />

**Reply by EngrPips** [reply]
@Equious sir

**Comment by DegenMouse** [comment]
I have just finished the Blockchain Fundamentals Course and it was truly a pleasure, and i felt like I really dipped my feet in blockchain. The videos were great and whenever the "give feedback page" appeared after the end of some videos I have always clicked the like button. But I think a small quality of life improvement would be really useful.

Right now when i click the "like" button you get this pop up only once:
<img width="412" alt="Screenshot 2025-04-09 at 13 07 58" src="https://github.com/user-attachments/assets/c0244766-8ce7-49e4-8d27-7a2a3c0c556d" />

But someone (i think it even happened to me) can accidentally click the wrong button, and there is no feedback to whether what button was pressed. So i think a on click highlight would be appreciated.

Something like this:
<img width="476" alt="Screenshot 2025-04-09 at 13 27 22" src="https://github.com/user-attachments/assets/c07e997c-02c9-4717-88b0-3eee71aadacc" />

**Reply by EngrPips** [reply]
Hello @SnavMaster09, Thanks for the feedback. @Equious sir

**Comment by leetebbs** [comment]
I am working on the Advanced Foundry , Develop an NFT Collection.
In section 15: Modify the NFT image onChain you use 

```
function flipMood(uint256 tokenId) public {
    if (!_isApprovedOrOwner(msg.sender, tokenId)) {
        revert MoodNft__NotOwnerNorApproved();
    }
    if (s_tokenIdToMood[tokenId] == Mood.HAPPY) {
        s_tokenIdToMood[tokenId] = Mood.SAD;
    } else {
        s_tokenIdToMood[tokenId] = Mood.HAPPY;
    }
}
```
but i cannot find _isApprovedOrOwner in the ERC721 Contract.

I have set mine to 

```
function flipMood(uint256 tokenId) public {
        if (ownerOf(tokenId) != msg.sender) {
            revert MoodNft__NotOwnerNorApproved();
        }
        if (s_tokenIdToMood[tokenId] == Mood.HAPPY) {
            s_tokenIdToMood[tokenId] = Mood.SAD;
        } else {
            s_tokenIdToMood[tokenId] = Mood.HAPPY;
        }
    }
```
Is this me or have things changed in the ERC721 openzepplin contract?

Great course by the way 
Thanks

**Reply by EngrPips** [reply]
Things changed, and these changes are always addressed in the chronological-update section https://github.com/Cyfrin/foundry-full-course-cu/blob/main/chronological-updates.md of the GitHub repo associated with the course.

**Comment by Damsel05** [comment]
Hi, Can I use my phone to do all these cos my laptop is bad at the moment.

**Reply by EngrPips** [reply]
You can sure try with your phone but a laptop will make things easier.

**Comment by Kelechikizito** [comment]
Hey @PatrickAlphaC, I appreciate you and your team (Cyfrin). The [Aragon DAO tutorial](https://updraft.cyfrin.io/courses/advanced-foundry/daos/introduction-to-aragon-dao) is currently outdated, so I thought I would bring it to your notice.

**Reply by EngrPips** [reply]
> Hey @PatrickAlphaC, I appreciate you and your team (Cyfrin). The [Aragon DAO tutorial](https://updraft.cyfrin.io/courses/advanced-foundry/daos/introduction-to-aragon-dao) is currently outdated, so I thought I would bring it to your notice.

Thanks for pointing this out @Kelechikizito. @PatrickAlphaC  @Equious

**Reply by ciaranightingale** [reply]
Hi @Kelechikizito thanks for flagging this - we are rectifying 👍 appreciate it ser

**Comment by mirmohmmadluqman** [comment]
Hi @PatrickAlphaC ! You have different written lessons and video lessons in Foundry Fundamentals, Raffle Lottery. In videos, you have a function namely `pickwinner()` but in written lessons there is `performupkeep()`..., and also in repo of it, there is whole **Raffle Lottery** with `performupkeep()`... function and you are teaching `pickwinner()` in videos.

**Reply by mirmohmmadluqman** [reply]
I had gave feedback to updraft, you have fixed in written lessons, but not repo yet

![WhatsApp Image 2025-09-28 at 20 53 02_3a6a9fe3](https://github.com/user-attachments/assets/e3396758-b38e-4702-be03-1598a9624c30)

**Reply by EngrPips** [reply]
@mirmohmmadluqman  Have you completed the video lesson for the `Raffle` section?

**Reply by mirmohmmadluqman** [reply]
@EngrPips Do you mean all videos !

**Reply by mirmohmmadluqman** [reply]
@EngrPips I know the name of function will be changed but I am telling that:
> In videos, you have a function namely pickwinner() but in written lessons there is performupkeep()...

They have fixed it now !

**Reply by EngrPips** [reply]
> @EngrPips I know the name of function will be changed but I am telling that:
> 
> > In videos, you have a function namely pickwinner() but in written lessons there is performupkeep()...
> 
> They have fixed it now !

Oh, Interesting. I am glad that has been fixed.

**Comment by mirmohmmadluqman** [comment]
I tried to do exam for it, and it showed me:

 > Oh no! Unfortunately, your score did not meet the minimum, and you did not pass the exam. The good news is that you can retake it in two days.
> Retake exam in *** * ** * *

---

and when I tried to give you feedback thrice even after reloading, it always showed 'Invalid Feedback' !

> <img width="294" height="114" alt="image" src="https://github.com/user-attachments/assets/078cd3e5-0154-47d6-adb9-b4742b082772" />

<img width="960" height="477" alt="image" src="https://github.com/user-attachments/assets/29cb7854-b938-4477-b443-4a9ab74a8eab" />

**Reply by EngrPips** [reply]
Please join the official Discord server of Cyfrin https://discord.gg/tZP4dE6C and report this issue using a `support ticket`

**Comment by mirmohmmadluqman** [comment]
It automatically answers the question without even touching any button

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/e3edbf16-ab1f-47c0-b8ba-2c5ae88008ce" />

**Reply by EngrPips** [reply]
@PatrickAlphaC @Equious

**Reply by EngrPips** [reply]
@mirmohmmadluqman, you may also use Cyfrin's official Discord to report this. Open a support ticket on there and report.

**Reply by Equious** [reply]
Thanks, we'll have a look.

**Comment by 1khushibarnwal** [comment]
Thank you so much @PatrickAlphaC for making such a wonderful course. 
I struggled a lot in the FundMe contract when you were explaining the Chainlink part, but I understand that's just a part of it. I would really appreciate if you could come up with more courses about BLOCKCHAIN ORACLES.

Once again, thanks for this.

Love from India.

**Reply by EngrPips** [reply]
Great Feedback.

**Comment by nargeslein** [comment]
Course Foundry Fundamentals [Lesson 1.3 Development environment setup (Mac, Linux)](https://updraft.cyfrin.io/courses/foundry/foundry-simple-storage/development-environment-setup-mac-linux)
I would suggest the following clarification:
- On Linux, the correct keyboard shortcut to clear the terminal is **Ctrl + L**.
- The shortcut **Ctrl + K** does not work in Linux terminals or in Visual Studio Code on Linux.
- The command **`clear`** works on all systems and is the safest option.

**Comment by nargeslein** [comment]
In [lesson 1.11 Memory Storage and Calldata](https://updraft.cyfrin.io/courses/solidity/simple-storage/solidity-memory-storage-calldata) of the course **Solidity Smart Contract Development**

### Clarification on “default to memory” wording
The written lesson contains has the following warning: 
> 🚧 WARNING:  
> Most variable types default to memory automatically. However, for strings, you must specify either memory or calldata due to the way arrays are handled in memory.

While discussing this with ChatGPT, it was pointed out that this wording could potentially be misleading.  
Primitive types like `uint`, `bool`, and `address` don’t actually default to `memory` and don’t have an explicit data location — the compiler decides whether they live on the stack, memory, or storage based on context.

The need to explicitly specify `memory` or `calldata` mainly applies to reference types such as `string`, arrays, and structs.

I’d be interested to hear if others agree with this interpretation or have additional insights.

**Reply by EngrPips** [reply]
This seems to make sense to me.

**Reply by PatrickAlphaC** [reply]
Good shout. They default to "the stack", and since solidity is a disaster of a compiler they should go into memory more often then they actually do. We will fix this!

**Reply by Equious** [reply]
<img width="1034" height="108" alt="image" src="https://github.com/user-attachments/assets/8e1af7ad-ce73-4f5c-964c-be2cca1d6620" />

**Reply by PatrickAlphaC** [reply]
Thank you!

**Comment by elenpoter394-sys** [comment]


**Reply by EngrPips** [reply]
@Equious Please take this down

**Comment by nargeslein** [comment]
In [lesson 4.5 Creating Custom Errors](https://updraft.cyfrin.io/courses/foundry/smart-contract-lottery/solidity-custom-errors) of the course **Foundry Fundamentals**, I noticed two elements that initially confused me but that I later clarified experimenting and dicussing with ChatGPT. I think explicitly mentioning these could help other students too.

## 1. Gas efficiency when comparing functions
The lesson compares gas usage between `require(string)` and custom errors, which makes perfect sense at this stage.

I experimented further and added multiple functions with identical logic but different names, reported gas costs can still differ. Additionally, adding another function, renaming a function, or changing their order can alter the gas numbers more than expected.

<img width="615" height="526" alt="Screenshot from 2026-02-07 15-59-56" src="https://github.com/user-attachments/assets/383dd78a-2821-42ff-803b-61264dfe004e" />

This led me to discover through ChatGPT that gas differences can come from **function selectors and dispatcher routing**, not just from the function body itself. In other words, identical logic does not always mean identical gas usage.

## 2. Contract layout vs. Solidity style guide
The lesson teaches being mindful of the contract layout when introducing custom errors, but I found it a bit confusing by comparing the official Solidity style guide with the commented code in the video.  

The [Solidity documentation](https://docs.soliditylang.org/en/latest/style-guide.html#order-of-layout) describes contract elements should be laid out in the following order:
- Pragma statements
- Import statements
- Events
- **Errors**
- Interfaces
- Libraries
- Contracts

While inside each contract, library or interface, the following is used:
- Type declarations
- State variables
 - Events
- **Errors**
- Modifiers
 - Functions

The video seems to show a "customized summary" of the two orders above in the comment code.
<img width="418" height="467" alt="Screenshot from 2026-02-07 15-48-16" src="https://github.com/user-attachments/assets/d0cff2d8-0d1b-4c96-ae3a-174cbe33fe27" />

The lesson is excellent and very informative. These are just small clarity points that could help students who like to experiment and reason more deeply about gas usage and contract structure. In general I noticed that **hands-on experimentation is a requirement to truly understand** beyond watching the videos.

**Reply by EngrPips** [reply]
Interesting, I didn't know initially that the gas cost for the routing function is included in the actual gas cost of the called function.

**Reply by nargeslein** [reply]
I asked ChatGPT 5.2 again (instead of the free version which I had consulted before) and I got a more details explanation:
<img width="852" height="362" alt="Screenshot from 2026-02-07 17-21-36" src="https://github.com/user-attachments/assets/51a2db1d-0e86-4966-8642-63a8e72ee625" />

**Reply by EngrPips** [reply]
Fair enough, that kind of makes sense to me, but it is just that I have never thought of it before.
